### ISSUE TITLE
IndexedDB connections must be closed so bfcache works

### ISSUE THREAD

**Issue by gustavopch** [firebase/firebase-js-sdk#6167]
### [REQUIRED] Describe your environment

  * Operating System version: macOS 12.3
  * Browser version: Chrome 100
  * Firebase SDK version: 9.6.7
  * Firebase Product: auth

### [REQUIRED] Describe the problem

As described in https://web.dev/bfcache, the browsers' back/forward cache doesn't work when IndexedDB has an open connection. Currently, connections are opened here and then they're never closed:

https://github.com/firebase/firebase-js-sdk/blob/cdada6c68f9740d13dd6674bcb658e28e68253b6/packages/auth/src/platform_browser/persistence/indexed_db.ts#L90-L94

They should be closed (`db.close()`) immediately after used or when `deleteApp()` is called.

**Comment by gustavopch** [comment]
In addition, when `deleteApp()` is called, `IndexedDBLocalPersistence` should stop polling IndexedDB.

Polling starts here:
https://github.com/firebase/firebase-js-sdk/blob/cdada6c68f9740d13dd6674bcb658e28e68253b6/packages/auth/src/platform_browser/persistence/indexed_db.ts#L402-L409

But then it keeps happening forever.

**Comment by gustavopch** [comment]
I'm experimenting here. Closing IndexedDB connections on `pagehide` event sometimes doesn't work. It seems the browser (Chrome v100 on macOS at least) doesn't give enough time before fully leaving the page. So closing them on `deleteApp` won't be enough to make bfcache work. Instead, it would be better to close the connections immediately after used. Pseudo-code:
```js
const runWithDatabase = async (fn) => {
  const db = await openDatabase()
  const result = await fn(db)
  closeDatabase(db)
  return result
}
```

**Comment by yoyomyo** [comment]
@gustavopch Is your experiment with a non-auth package, as suggested in https://github.com/firebase/firebase-js-sdk/issues/6167#issuecomment-1108908079?

**Comment by gustavopch** [comment]
@yoyomyo I've tested Auth and Firestore. The IndexedDB issue explained in this issue only seems to happen with Auth.

There's also something in Firestore that breaks bfcache under certain circumstances, but I didn't experiment enough to bring the reproduction steps.

**Comment by gustavopch** [comment]
Solving this issue may involve refactoring `deleteApp()` to be synchronous. If it can synchronously clean up all IndexedDB connections, web socket subscriptions and HTTP requests, having `window.addEventListener('pagehide', () => deleteApp(getApp()))` in user's code may be enough to make bfcache work.

**Comment by gustavopch** [comment]
On Safari for iOS v15.5, bfcache will sort of work regardless of closing IndexedDB connections or running `deleteApp()`, but there's a big problem: if you call `deleteApp()` on `pagehide`, when you come back to that page and it's restored from bfcache, a newly initialized Firestore instance will just not work anymore. Probably due to an incomplete cleanup. This is a serious bug and I think it should elevate the priority of this issue.

**Comment by Whip** [comment]
Its nearly 3 years since the issue was opened. Why do Google devs code so poorly and take years to fix their mess? Please take this on priority.

**Comment by gustavopch** [comment]
@Whip Yeah... look at [this PR](https://github.com/firebase/firebase-tools/pull/6191). It's a one-liner that was [requested](https://github.com/firebase/firebase-tools/issues/6112#issuecomment-1652539666) by a team member, yet it's been overlooked for almost two years. If they can't review and merge a simple, one-line change, what can we really expect? Honestly, I hope that one day Google will merge Firebase into Google Cloud so it can benefit from a larger team. In the meantime, I've discovered Convex and it fulfills everything I wish Firebase would (there's also Supabase, Appwrite, etc.).
